### 📊 Nigeria Financial Market Dataset Analysis

#### Overview
The Nigeria Financial Exchange Market dataset downloaded from CBN site contains financial market data capturing exchange rates and turnover activity across different trading segments. It provides insight into market liquidity, pricing trends, and trading volume dynamics.

#### Key Variables
- **ratedate**: The date of the recorded market data
- **closingrate**: The closing exchange rate for the day
- **nfeM_Total_Turnover**: Total turnover in the NFEM market segment
- **interBank_Total_Turnover**: Total turnover in the interbank market

#### Objective
The goal of this analysis is to:
- Examine trends in exchange rates over time  
- Analyze liquidity patterns using turnover data  
- Identify anomalies or significant movements  
- Generate actionable financial insights  

#### Scope
This analysis focuses on:
- Time-based trends  
- Market activity comparison  
- Statistical summaries  
- Insight generation using AI (Qwen model)

In [1]:
# import packages
import pandas as pd
import numpy as np

In [2]:
# import dataset
df = pd.read_csv(r"C:\Users\ACER\Desktop\Portfolios\NFEM\NFEM_data.csv")
# show first few rows
df.head()

,ratedate,closingrate,highestrate,lowestrate,weightedAvgRate,simpleAvgRate,noOfDeals,noOfDeals_InterBank,nfeM_Total_Turnover,interBank_Total_Turnover
0,3/27/2026,1384.25,1384.5,1377.0,1380.5766,1381.0235,0,141,0.0,1.728924e+08
1,3/26/2026,1389.00,1391.0,1375.0,1383.8830,1385.5284,318,107,375654114.3,9.152994e+07
2,3/25/2026,1391.00,1391.5,1376.0,1386.7049,1387.2227,231,64,325962381.4,5.569755e+07
3,3/24/2026,1383.50,1389.0,1372.0,1382.6325,1381.8642,293,88,480096825.8,8.343844e+07
4,3/23/2026,1383.00,1399.0,1380.0,1388.3841,1387.7055,404,144,984139865.8,1.881261e+08


In [5]:
# transform data
df["ratedate"] = pd.to_datetime(df["ratedate"])
df = df.sort_values("ratedate")
df.head()


,ratedate,closingrate,highestrate,lowestrate,weightedAvgRate,simpleAvgRate,noOfDeals,noOfDeals_InterBank,nfeM_Total_Turnover,interBank_Total_Turnover
331,2024-12-02,1660.0,1678.0,1650.0,1661.1225,1662.7719,0,0,0.0,0.0
330,2024-12-03,1625.0,1664.0,1623.0,1643.1507,1644.7799,0,0,0.0,0.0
329,2024-12-04,1608.0,1630.0,1590.0,1613.6946,1613.8600,0,0,0.0,0.0
328,2024-12-05,1567.0,1610.0,1565.0,1588.7939,1587.2944,0,0,0.0,0.0
327,2024-12-06,1535.0,1575.0,1510.0,1533.7598,1533.9388,0,0,0.0,0.0


In [5]:
# create key financial metrics
df["total_turnover"] = df["nfeM_Total_Turnover"] + df["interBank_Total_Turnover"]
df.head(10)


,ratedate,closingrate,highestrate,lowestrate,weightedAvgRate,simpleAvgRate,noOfDeals,noOfDeals_InterBank,nfeM_Total_Turnover,interBank_Total_Turnover,total_turnover
0,3/27/2026,1384.25,1384.50,1377.0,1380.5766,1381.0235,0,141,0.000000e+00,1.728924e+08,1.728924e+08
1,3/26/2026,1389.00,1391.00,1375.0,1383.8830,1385.5284,318,107,3.756541e+08,9.152994e+07,4.671841e+08
2,3/25/2026,1391.00,1391.50,1376.0,1386.7049,1387.2227,231,64,3.259624e+08,5.569755e+07,3.816599e+08
3,3/24/2026,1383.50,1389.00,1372.0,1382.6325,1381.8642,293,88,4.800968e+08,8.343844e+07,5.635353e+08
4,3/23/2026,1383.00,1399.00,1380.0,1388.3841,1387.7055,404,144,9.841399e+08,1.881261e+08,1.172266e+09
5,3/18/2026,1362.00,1362.00,1349.0,1353.9018,1355.0659,255,86,2.427062e+08,7.244048e+07,3.151467e+08
6,3/17/2026,1345.00,1355.00,1340.0,1344.4158,1344.5358,342,136,3.709795e+08,1.044806e+08,4.754601e+08
7,3/16/2026,1355.00,1365.35,1354.0,1357.7697,1358.7150,272,88,3.462756e+08,8.575386e+07,4.320294e+08
8,3/13/2026,1363.50,1374.00,1362.0,1366.2330,1366.4089,374,146,1.127762e+09,1.799118e+08,1.307674e+09
9,3/12/2026,1370.00,1390.00,1362.0,1371.5063,1372.6543,354,142,7.004353e+08,1.596985e+08,8.601338e+08


In [7]:
# basic statistical analysis
df[["closingrate", "total_turnover"]].describe()

,closingrate,total_turnover
count,332.000000,3.320000e+02
mean,1498.145392,2.605052e+07
std,70.540641,1.551547e+08
min,1337.000000,0.000000e+00
25%,1448.625000,0.000000e+00
50%,1513.000000,0.000000e+00
75%,1540.000000,0.000000e+00
max,1660.000000,1.347918e+09


In [13]:
# trend analysis
df.tail(5)

,ratedate,closingrate,highestrate,lowestrate,weightedAvgRate,simpleAvgRate,noOfDeals,noOfDeals_InterBank,nfeM_Total_Turnover,interBank_Total_Turnover,total_turnover
4,2026-03-23,1383.00,1399.0,1380.0,1388.3841,1387.7055,404,144,984139865.8,1.881261e+08,1.172266e+09
3,2026-03-24,1383.50,1389.0,1372.0,1382.6325,1381.8642,293,88,480096825.8,8.343844e+07,5.635353e+08
2,2026-03-25,1391.00,1391.5,1376.0,1386.7049,1387.2227,231,64,325962381.4,5.569755e+07,3.816599e+08
1,2026-03-26,1389.00,1391.0,1375.0,1383.8830,1385.5284,318,107,375654114.3,9.152994e+07,4.671841e+08
0,2026-03-27,1384.25,1384.5,1377.0,1380.5766,1381.0235,0,141,0.0,1.728924e+08,1.728924e+08


In [15]:
# volatility (rate changes)
df["rate_change"] = df["closingrate"].diff()
df[["ratedate", "closingrate", "rate_change"]].tail(10)

,ratedate,closingrate,rate_change
9,2026-03-12,1370.00,-3.50
8,2026-03-13,1363.50,-6.50
7,2026-03-16,1355.00,-8.50
6,2026-03-17,1345.00,-10.00
5,2026-03-18,1362.00,17.00
4,2026-03-23,1383.00,21.00
3,2026-03-24,1383.50,0.50
2,2026-03-25,1391.00,7.50
1,2026-03-26,1389.00,-2.00
0,2026-03-27,1384.25,-4.75


In [ ]:
import requests

summary_text = summary.to_string()

prompt = f"""
You are a financial analyst.

Analyze this dataset summary and provide:
1. Key trends
2. Risk insights
3. Business recommendations

Data:
{summary_text}
"""

response = requests.post(
    "http://localhost:11434/api/generate",
    json={
        "model": "qwen2.5:7b",
        "prompt": prompt,
        "stream": False
    }
)

print(response.json()["response"])

### Analysis

#### Key Trends in Exchange Rates

1. **Mean and Range**:
   - The mean closing rate is approximately 1498.15, with a standard deviation of about 70.54.
   - The exchange rate ranges from 1337 to 1660, indicating significant volatility.

2. **Recent Trends**:
   - In the recent data, the exchange rates have shown some fluctuations around 1380-1390, with a slight downward trend over the last few days.
   - Notable changes include:
     - March 23: Rate = 1383.00
     - March 24: Rate = 1383.50 (down 0.50%)
     - March 25: Rate = 1391.00 (up 6.71%)
     - March 26: Rate = 1389.00 (down 1.44%)
     - March 27: Rate = 1384.25 (down 3.06%)

#### Liquidity Insights from Turnover

1. **Turnover Distribution**:
   - The minimum turnover is zero, and the maximum is 1.347918e+09.
   - Most turnovers are around 3-10 billion units.

2. **Liquidity Observations**:
   - There have been days with no trading activity (e.g., March 27), indicating periods of low liquidity.
   - Days like March 25, which had a high turnover of approximately 381.66 billion units, suggest periods of high market interest and activity.

#### Risk Observations

1. **Volatility**:
   - The exchange rate has shown significant volatility with daily changes ranging from -4.75% to +6.71%.

2. **Risk Factors**:
   - High turnover on certain days can indicate higher liquidity but also increased market volatility, which may pose risks for investors.
   - Days with zero trading activity suggest periods of low liquidity and potential market instability.

#### Clear Business or Investment Recommendations

1. **Diversify Exposure**:
   - Given the high volatility and risk observed in recent data, it is advisable to diversify exposure across different timeframes and instruments to mitigate risks.

2. **Monitor Liquidity Levels**:
   - Keep a close watch on days with low trading volumes (e.g., March 27) as these can indicate periods of market stress or regulatory changes.

3. **Consider Hedging Strategies**:
   - Implement hedging strategies to protect against sudden rate movements, especially given the observed volatility.

4. **Strategic Positioning**:
   - Consider short-term trading strategies on days with higher turnover and lower volatility.
   - Maintain a long-term view but be prepared for market corrections or reversals.

5. **Review Regulatory Changes**:
   - Monitor any recent regulatory changes that could impact exchange rates and liquidity, as they can significantly affect market behavior.

By following these recommendations, investors can better navigate the uncertainties in the exchange rate market and manage their risks effectively.

In [20]:
import os
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer
from reportlab.lib.styles import getSampleStyleSheet
from reportlab.lib.pagesizes import letter

def create_pdf(summary_stats, ai_text):
    # 1. Define the exact folder path
    folder = r"C:\Users\ACER\Desktop\Portfolios\NFEM"
    
    # Create folder if it doesn't exist
    if not os.path.exists(folder):
        os.makedirs(folder)

    file_path = os.path.join(folder, "financial_report.pdf")
    print(f"Attempting to save PDF to: {file_path}")

    # 2. Setup Document
    doc = SimpleDocTemplate(file_path, pagesize=letter)
    styles = getSampleStyleSheet()
    story = []

    # 3. Add Content
    story.append(Paragraph("FOREIGN EXCHANGE MARKET ANALYSIS REPORT", styles["Title"]))
    story.append(Spacer(1, 12))

    story.append(Paragraph("Executive Summary", styles["Heading2"]))
    
    # Use .get() or direct keys from your .describe() output
    stats_text = (
        f"Average Rate: {summary_stats['closingrate']['mean']:.2f}<br/>"
        f"Max Rate: {summary_stats['closingrate']['max']:.2f}<br/>"
        f"Min Rate: {summary_stats['closingrate']['min']:.2f}<br/>"
        f"Volatility (Std Dev): {summary_stats['closingrate']['std']:.2f}"
    )
    story.append(Paragraph(stats_text, styles["Normal"]))
    story.append(Spacer(1, 12))

    # 4. Add AI Insights
    story.append(Paragraph("Market Analysis & Insights", styles["Heading2"]))
    # Cleaning up text for ReportLab compatibility
    clean_ai_text = ai_text.replace("\n", "<br/>")
    story.append(Paragraph(clean_ai_text, styles["Normal"]))

    # 5. Build PDF
    try:
        doc.build(story)
        print("PDF successfully created in NFEM folder ✅")
    except Exception as e:
        print(f"An error occurred: {e}")

# --- EXECUTION BLOCK ---
# This actually triggers the function using your variables
stats = df[["closingrate", "total_turnover"]].describe().to_dict()
# Extracting the markdown text you wrote in the notebook as a string
insights = """The exchange rate ranges from 1337 to 1660, indicating significant volatility. 
Recent trends show fluctuations around 1380-1390. 
Liquidity observations suggest periods of high market interest followed by zero-trading days."""

create_pdf(stats, insights)

Attempting to save PDF to: C:\Users\ACER\Desktop\Portfolios\NFEM\financial_report.pdf
PDF successfully created in NFEM folder ✅
